server/api/classyear/classyear.controller.js

```diff
@@ -220,5 +222,20 @@ function generateCode(codeLength){
       var character = (Math.floor(Math.random() * characterOptions.length));
       code = code.concat(characterOptions[character.toString()]);
   }
+  ClassYear.findOne({"dayCodes.code":code})
```

Comment:
This function is expected to be synchronous. By adding in the ClassYear.findOne, you are making it async. Anything within the findOne block can no longer correctly return a string to the calling function, and it will break. Also, by having a `handleError(res, err);` you are both trying to return an error and using variables that have not been defined in the function. Try hardcoding in a code that you know is in the database (for the top level of recursion only) to try out your code.  In the long term, you should probably write a async function that handles calling generateCode as many times as needed to get a unique one. I would also suggest that if you get a collision, increase the size of the code by 1 to avoid the possibility of failing many times in a row.